lib/node_modules/@stdlib/array/base/join/lib/main.js

```diff
@@ -0,0 +1,136 @@
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var arraylike2object = require( '@stdlib/array/base/arraylike2object' );
+
+
+// FUNCTIONS //
+
+/**
+* Tests whether an object has a specified method.
+*
+* @private
+* @param {Object} obj - input object
+* @param {string} method - method name
+* @returns {boolean} boolean indicating whether an object has a specified method
+*
+* @example
+* var bool = hasMethod( [], 'join' );
+* // returns true
+*
+* @example
+* var bool = hasMethod( [], 'beep' );
+* // returns false
+*/
+function hasMethod( obj, method ) {
+	return ( typeof obj[ method ] === 'function' );
+}
+
+/**
+* Returns a string which is the concatenation of all the elements in an array with a given separator when input is an accesor array.
+*
+* @private
+* @param {Object} x - input array object
+* @param {integer} separator - separator
+* @returns {string} output array
+*
+* @example
+* var toAccessorArray = require( '@stdlib/array/base/to-accessor-array' );
+* var arraylike2object = require( '@stdlib/array/base/arraylike2object' );
+*
+* var x = arraylike2object( toAccessorArray( [ 1, 2, 3, 4 ] ) );
+*
+* var out = accessors( x, ',' );
+* // returns '1,2,3,4'
+*/
+function accessors( x, separator ) {
+	var output;
+	var data;
+	var get;
+	var i;
+	data = x.data;
+	get = x.accessors[ 0 ];
+	output = '';
+	for ( i = 0; i < data.length; i++ ) {
+		if ( typeof get( data, i ) === 'undefined' || get( data, i ) === null ) {
+			output += separator;
+			continue;
+		}
+		output += get( data, i );
+		if ( i < data.length - 1 ) {
+			output += separator;
+		}
+	}
+	return output;
+}
+
+
+// MAIN //
+
+/**
+* Returns a string which is the concatenation of all the elements in an array with a given separator.
+*
+* @param {Collection} x - input array
+* @param {integer} separator - separator to be used in string
+* @returns {string} output string
+*
+* @example
+* var x = [ 1, 2, 3, 4 ];
+*
+* var out = join( x, ',' );
+* // returns '1,2,3,4'
+*
+* @example
+* var x = [ 1, 2, 3, null, undefined,  4 ];
+*
+* var out = join( x, '-' );
+* // returns '1-2-3---4'
+*/
+function join( x, separator ) {
+	var obj;
+	var i;
+	var s;
+	s='';
+	if ( hasMethod( x, 'join' ) ) {
+		return x.join( separator );
+	}
+	obj = arraylike2object( x );
+	if ( obj.accessorProtocol ) {
+		return accessors( obj, separator );
+	}
+	for ( i = 0; i < x.length; i++ ) {
+		if ( typeof x[ i ] === 'undefined' || x[ i ] === null ) {
+			s += separator;
```

Comment:
Same comments as above. Also, cache the value. If `x` is proxied, this can result in significantly degraded performance.